Pass *-options to GHC when compiling ordinary Haskell sources

zlonast opened the pull request:
Will close issues https://github.com/haskell/cabal/issues/9801 https://github.com/haskell/cabal/issues/4435
Main idea https://github.com/haskell/cabal/issues/9801#issuecomment-2884201996
> IMHO, cc-options (and ld-options and cpp-options and ...) should be always passed when invoking GHC, similarly as ghc-options should be always used when invoking ghc - regardless of what is the intention of a particular GHC-call. GHC might use or not use the options, Cabal cannot know and should not guess.

`cc-options`, `cxx-options`, `jspp-options`, `asm-options`, `cmm-options`, `ld-options`, `cpp-options`, 
should be always passed when invoking GHC, similarly as `ghc-options` should be always used when
invoking `ghc` - regardless of what is the intention of a particular GHC-call.
GHC might use or not use the options, Cabal cannot know and should not guess.

`cc-options` and `cxx-options` still need to be separated (C/C++) for versions below 8.10.
https://gitlab.haskell.org/ghc/ghc/-/issues/16477

---

**Template Α: This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
  * [x] [Is the change significant?](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#is-my-change-significant) If so, remember to add `significance: significant` in the changelog file.
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

---

Depends-on: #11711
Depends-on: #11713

zlonast commented:
@ulysses4ever @phadej It's not very easy to add flags to sources files due to backward compatibility, so for now I suggest considering the pull request as is ¯\\\_(ツ)_/¯ (Anyway, this solved the problem with macros in header files)

ulysses4ever reviewed (commented):
Great start, thank you! Below are some inline questions.

One general question: did you check that this doesn't lead to duplicated options when cabal calls GHC? I assume other *-options are passed today sometimes. So, sometimes we'll get those options passed as today and through your change too. Is that right? It seems undesirable.

> It's not very easy to add flags to sources files due to backward compatibility

Can you be more specific? Do you mean it's hard to test on older GHCs? This shouldn't be a problem because the tests can run for specific versions of GHC (e.g. starting from 8.10).

geekosaur reviewed (changes requested):
Same concerns as @ulysses4ever raised.

phadej commented on Cabal/src/Distribution/Simple/GHC/Internal.hs:
```diff
@@ -377,6 +378,18 @@ includePaths lbi bi clbi odir =
+    ++ ["-optJSP" ++ opt | opt <- jsppOptions bi]
+    ++ ["-optl" ++ opt | opt <- ldOptions bi]
+    ++ ["-optc" ++ opt | opt <- ccOptions bi]
+#if __GLASGOW_HASKELL__ >= 810
```
this is wrong. You should check for the configured compiler, not the one used to compile `Cabal`.

zlonast commented on cabal-testsuite/PackageTests/FFI/ForeignOptsCapi/README.md:
```diff
@@ -0,0 +1,3 @@
+# ForeignOptsCapi
+
+This test case asserts that cabal passes `cc-options` to the C compiler when use `foreign import capi`.
```
I need to think about it, but I don't know yet. Maybe I need to read other issues to see if anything worked.

ulysses4ever replied:
Could you, at least, check that the test suite has tests for all the other-language options (ASM, JS, CXX, CMM, LD)? And if some of these options don't have tests, could you add something simple for them? I agree that it's not fun to do this exercise, but I think it's a big change and I want at least a minimal guarantee that we're not breaking things badly.

zlonast replied:
That's a really good opinion, I'll check it out and come back with the results.

zlonast replied:
`cc-options` - `cabal-testsuite/PackageTests/FFI/ForeignOptsC/README.md`
`cxx-options` - `cabal-testsuite/PackageTests/FFI/ForeignOptsCxx/README.md`
`jspp-options` - `cabal-testsuite/PackageTests/JS/JSPPOptions/README.md`
`asm-options` - none
`cmm-options` - `cabal-testsuite/PackageTests/Cmm/CmmSources/README.md`
`ld-options` - `ParserTests`
`cpp-options` - `ParserTests`
`-pgmc` - `cabal-testsuite/PackageTests/ShowBuildInfo/Complex`
`-O1` and `-O2` for C++ and C and Asm - `cabal-testsuite/PackageTests/ShowBuildInfo/Complex`

zlonast commented on changelog.d/pr-10969.md:
```diff
@@ -0,0 +1,10 @@
+prs: 10969
+---
+
+*-options should be always passed when invoking GHC,
```
They were only partially in `componentGhcOptions`, after my changes, some cross-interaction became possible. This is currently only available for `.c` and `.h` or `.cpp/.cxx/.c++` and `.h`. I think so.

zlonast replied:
@ulysses4ever Perhaps I answered your question?

ulysses4ever replied:
Yes, thank you!

zlonast commented on Cabal/src/Distribution/Simple/GHC/Internal.hs:
```diff
@@ -377,142 +376,41 @@ includePaths lbi bi clbi odir =
+optimizationCFLags lbi =
+  ( case withOptimization lbi of
+      NoOptimisation -> []
+      NormalOptimisation -> ["-O1"] -- _ -> ["-O2"] -- attention: needs-review toGhcOptimisation
```
I'm not sure about this change, even though it seems logical.

zlonast commented:
> One general question: did you check that this doesn't lead to duplicated options when cabal calls GHC? I assume other *-options are passed today sometimes. So, sometimes we'll get those options passed as today and through your change too. Is that right? It seems undesirable.

@ulysses4ever  Yes, you're right, now I've fixed it.

> Can you be more specific? Do you mean it's hard to test on older GHCs? This shouldn't be a problem because the tests can run for specific versions of GHC (e.g. starting from 8.10).

@phadej correctly noted that `compilerCompatVersion` should have been added.

> I'd say the problem is rather that GhcOptions is not used uniformly. There is e.g. componentCcGhcOptions which does set ghcOptCcOptions. Why these don't show up when compiling files CApiFFI headers?

Because they were only transferred to `buildAllExtraSources` files.

> IMHO, it's wrong that there are separate componentAsmGhcOptions, componentJsGhcOptions, componentGhcOptions. There should be just one componentGhcOptions with all the options, and then it's up to GHC which ones are relevant, and which aren't.

I think you're right, now I did it this way.

> EDIT: There might be situations where ghcOptMode or something like that is different, and input files "obviously". But the bulk of options should be the same. 

Yes, I separated them into `sourcesGhcOptions` and `componentGhcOptions`.

> Currently there are six copies of that. Cannot be right. I suspect there are many subtle inconsistencies hiding because of that.

Looks like I'll have to think about how to break the `cabal` :)

> EDIT: So AFAICT, this is not the correct fix, not what I'd imagine seeing.

I hope you like the new version better.

geekosaur commented on changelog.d/pr-10969.md:
```diff
@@ -0,0 +1,10 @@
+---
+synopsis: Pass *-options to GHC when compiling ordinary Haskell sources
+packages: [Cabal]
+prs: 10969
```
Now that it's been added, either you or Artem should mark this resolved.

phadej commented on Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs:
```diff
@@ -86,7 +102,21 @@ buildCSources mbMainFile =
+          { ghcOptCcOptions =
+              ( case compilerCompatVersion GHC (compiler lbi) of
+                  Just v
+                    | v >= mkVersion [8, 10] -> Internal.defaultGhcOptCcOptions lbi bi
```
why this is not in `sourcesGhcOptions `. It has access to `lbi`?

zlonast replied:
It does not completely match, there are different `defaultGhcOptCcOptions` and `defaultGhcOptCxxOptions`.
You are right, `ghcOptCcOptions` and `ghcOptCxxOptions` can be moved to `sourcesGhcOptions`.
On the other hand, `ghcOptCcProgram` should be left here, since not everywhere (it seems to me) there may be access to `gccProgram`.

phadej replied:
> here may be access to gccProgram.

That's a bug on its own:
- If `--with-gcc` is given, it should always be passed to GHC (so GHC always consistently uses the same GCC).

phadej replied:
FWIW, here in `buildCxxSources ` you use `defaultGhcOptCcOptions `, and in `buildCSources` you set `defaultGhcOptCxxOptions `.

That's just proves my point, subtle differences potentially leading to bugs.

zlonast replied:
> I only see a downsides of using subtly different calls to GHC when compiling C or C++ files.

I don't understand, in 8.10 there was a flag split, cabal decided to support versions older than 8.10 too. Maybe you think it's a bug?

> That's just proves my point, subtle differences potentially leading to bugs.

This is because we maintain the same behavior for versions older than 8.10.

zlonast replied:
If you think this is a bug, then we need to remove this test.

`PackageTests/FFI/ForeignOptsCxx`

phadej replied:
And arguably we should check for supported GHC version in one place.

I'm not sure where it's done now, at GhcOptions creation time or its serialisation time.

Either way, if some options are dropped because of GHC not supporting them, they shouldn't be dropped silently.

phadej replied:
No. Am I explaining myself unclearly. Let's try again.

it looks like that

- You set C options and `ghcOptCcProgram` **only** in `buildCSources`
- You set C++ options and `ghcOptCxxProgram` **only** in `buildCxxSources`

That cannot be right. You should set these options **always**.

phadej replied:
In the retrospective, `cabal-install` could refused to support an accidental "but it did work before with older GHC, cc-options were passed to C compiler" when proper `cxx-options` were added. Those accidents just make code base very spagetti.

zlonast replied:
It seems I have implemented everything that is written here? With reservations for versions below 9.4.

zlonast commented:
> @zlonast: so are the CI failures expected? Some at least look intermittent (Internet is aging), so let me take the liberty of rebasing on master where some tests are marked flaky. Maybe it helps.
> 
> Regardless, if anybody would like to help move this PR forward, in particular answer the last question about testing, that would be very helpful. Thank you!

Yes, thanks, now I'm thinking of fixing the test for Windows (ignoring the different paths to the `-pgmc` file) and porting the options from Bazel for static libs (or as an alternative, not passing the `--with-gcc` flag for versions below 9.4.

https://github.com/tweag/rules_haskell/blob/3322610caba2ce8f17ec010b2f76a366a06f97c4/haskell/cabal.bzl#L321-L329

zlonast commented:
> > @Mikolaj can we add `GCCPATH: ${{ steps.setup-haskell.outputs.gcc-exe }}`? I'm not very good at writing tests where the path for `-pgmc` appears
> 
> @zlonast: do you mean to cabal CI? What does it do?

It's like GHCPATH just for gcc or clang, now I wrote 5 tests because of different paths in different windows versions :(

Mikolaj commented:
Sounds good to me. Please go ahead. But just in case and let me ping @geekosaur, @ulysses4ever and @mpickering about this change and the review of the whole PR.

ulysses4ever commented:
@zlonast can you comment which suggestions were implemented and give an overview of what's done currently?

zlonast commented:
> @zlonast can you comment which suggestions were implemented and give an overview of what's done currently?

In this pr I added: `cc-options`, `cxx-options`, `jspp-options`, `asm-options`, `cmm-options`, `ld-options`, `cpp-options` and `ccProgram` to `componentGhcOptions`.

Removed all `componentCcGhcOptions`, `componentCxxGhcOptions`, `componentAsmGhcOptions`, `componentJsGhcOptions`, `componentCmmGhcOptions`.

These options still need to be separated for older versions.
`cc-options` and `cxx-options` need `separateGhcOptions` for versions below 8.10.

The suggestion that gcc(`ccProgram`) can always be passed is good, but this is not entirely true for versions below 9.4.
Because when passing `ghc` implicitly, it added the correct flags itself. (probably it is possible to read it somehow, but I lost, see the message above)

I also took out the common part of the linking options.
`componentGhcOptions` and `linkGhcOptions`

Written a test for `capi`.
Written tests for `ShowBuildInfo` (there are 5 of them now, but if we add GCCPATH, we can make 2).

ulysses4ever commented on Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs:
```diff
@@ -73,7 +74,15 @@ buildCSources
+        (Internal.sourcesGhcOptions verbosity lbi bi clbi odir filename)
+          { -- C++ compiler options: GHC >= 8.10 requires -optcxx, older requires -optc
+            -- we want to be able to support cxx-options and cc-options separately
+            ghcOptCxxOptions = Internal.separateGhcOptions (mkVersion [8, 10]) (compiler lbi) (Internal.defaultGhcOptCxxOptions lbi bi)
```
This is probably silly but I'll ask anyway: why do we touch a cxx-related field in a function that's called `buildCSources`? Looks like only C-related options should be touched. 

I did read the comment above, which probably explains it, but I didn't understand how it's related to my conccern.

zlonast replied:
hmm, I guess I should somehow refer to the tests `cabal-testsuite/PackageTests/FFI/ForeignOptsC`

```C
#include "clib.h"

#ifndef __TESTOPT_C__
#error "Did not get required __TESTOPT_C__ from cc-options"
#endif

#ifdef __TESTOPT_CXX__
#error "Got unexpected __TESTOPT_CXX__ from cxx-options"
#endif

int meaning_of_life_c() {
    return __TESTOPT_C__;
}
```
the test is just for this
```
executable foreign-opts-c-exe
  main-is:             Main.hs
  build-depends:       base
  default-language:    Haskell2010
  include-dirs:        cbits
  c-sources:           cbits/clib.c
  -- The following options are shared in all ForeignOpts cases:
  cc-options:          -D__TESTOPT_C__=11
  cxx-options:         -D__TESTOPT_CXX__=22
```

ulysses4ever commented on Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs:
```diff
@@ -73,7 +74,15 @@ buildCSources
+          { -- C++ compiler options: GHC >= 8.10 requires -optcxx, older requires -optc
+            -- we want to be able to support cxx-options and cc-options separately
+            ghcOptCxxOptions = Internal.separateGhcOptions (mkVersion [8, 10]) (compiler lbi) (Internal.defaultGhcOptCxxOptions lbi bi)
+          , -- there are problems with linking with versions lower than 9.4, that's why we need this duplication
```
what "duplication"? With the code for `buildCxxOptions` below? Please, mention it

zlonast replied:
I need to change the description, then the line is needed to re-assign `ghcOptCcProgram`

ulysses4ever commented on Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs:
```diff
@@ -86,7 +95,15 @@ buildCSources mbMainFile =
+        (Internal.sourcesGhcOptions verbosity lbi bi clbi odir filename)
+          { -- C++ compiler options: GHC >= 8.10 requires -optcxx, older requires -optc
+            -- we want to be able to support cxx-options and cc-options separately
+            ghcOptCcOptions = Internal.separateGhcOptions (mkVersion [8, 10]) (compiler lbi) (Internal.defaultGhcOptCcOptions lbi bi)
+          , -- there are problems with linking with versions lower than 9.4, that's why we need this duplication
+            ghcOptCcProgram = Internal.defaultGhcOptCcProgram lbi
```
Same: C-options touched inside a Cxx-function. If this is related to the optcxx vs optc mishmash pre 8.10, I need some sort of simple example to understand how what's done here (and above) helps with that mishmash.

zlonast replied:
I should somehow refer to the tests `cabal-testsuite/PackageTests/FFI/ForeignOptsCxx`

ulysses4ever commented on Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs:
```diff
@@ -211,14 +226,12 @@ buildExtraSources
                 vanillaSrcOpts
                   `mappend` mempty
                     { ghcOptProfilingMode = toFlag True
-                    , ghcOptFPic = toFlag True
```
The change to fpic business is worrying, it should be loudly documented (including the changelog, probably).

zlonast replied:
This is a duplicate of the `vanillaSrcOpts`

```haskell
              vanillaSrcOpts =
                -- -fPIC is used in case you are using the repl
                -- of a dynamically linked GHC
                baseSrcOpts{ghcOptFPic = toFlag True}
```

ulysses4ever commented on Cabal/src/Distribution/Simple/GHCJS.hs:
```diff
@@ -1436,13 +1435,17 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
-                clbi
-                tmpDir
-                filename
+              ( Internal.sourcesGhcOptions verbosity lbi bnfo clbi tmpDir filename
+                  <> mempty
+                    { ghcOptCcOptions =
```
```haskell
              ( Internal.sourcesGhcOptions verbosity lbi bnfo clbi tmpDir filename
                  <> mempty
                    { ghcOptCcOptions =
```
is it not the same as
```suggestion
              ( (Internal.sourcesGhcOptions verbosity lbi bnfo clbi tmpDir filename)
                  { ghcOptCcOptions =
```
? (I hope I'm not making up this syntax!)

zlonast replied:
Yes, you are right, it should be done the same for GHC and GHCJS, it should be done without `mempty`, because it is a replacement, not an addition.

ulysses4ever commented on changelog.d/pr-10969.md:
This should have `significant: true`. I have a slight suspicion that we'll get a fallout after this change, so I want it to be extra visible.

ulysses4ever reviewed (commented):
So, what about possible option duplication? Are we sure we don't pass things twice? Even on older GHCs?

zlonast commented on Cabal/src/Distribution/Simple/GHC.hs:
```diff
@@ -744,7 +743,7 @@ libAbiHash verbosity _pkg_descr lbi lib clbi = do
           , ghcOptHiSuffix = toFlag "dyn_hi"
           , ghcOptObjSuffix = toFlag "dyn_o"
-          , ghcOptExtra = hcOptions GHC libBi ++ hcSharedOptions GHC libBi
+          , ghcOptExtra = hcSharedOptions GHC libBi
```
This is a duplicate of options from `componentGhcOptions`

```haskell
    vanillaArgs =
      (Internal.componentGhcOptions verbosity lbi libBi clbi (componentBuildDir lbi clbi))
        `mappend` mempty
          { ghcOptMode = toFlag GhcModeAbiHash
          , ghcOptInputModules = toNubListR $ exposedModules lib
          }
    sharedArgs =
      vanillaArgs
        `mappend` mempty
          { ghcOptDynLinkMode = toFlag GhcDynamicOnly
          , ghcOptFPic = toFlag True
          , ghcOptHiSuffix = toFlag "dyn_hi"
          , ghcOptObjSuffix = toFlag "dyn_o"
          , ghcOptExtra = hcOptions GHC libBi ++ hcSharedOptions GHC libBi
          }
```

mpickering commented:
What is the testing strategy for this patch? This is some of the most sensitive code in `cabal`, so I am nervous that many lines are changed in this patch.

mpickering reviewed (commented):
Can you explain these comments a bit more, I don't understand what "separated" means.

```

`cc-options` and `cxx-options` still need to be separated (C/C++) for versions below 8.10.

gccProgram till need to be separated (C/C++) for versions below 9.4.
```

I can appreciate that this change is right direction of travel but also requires very much care since there's a high likelihood of either a bug or a change in the behaviour when a user program is compiled.

mpickering commented:
I compiled stackage with this branch and GHC-9.10.2 and it all seemed to compile.

zlonast commented on Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs:
```diff
@@ -86,7 +102,22 @@ buildCSources mbMainFile =
+                (mkVersion [8, 10])
+                (compiler lbi)
+                (Internal.defaultGhcOptCcOptions lbi bi)
+          , -- there are problems with linking with versions below 9.4,
```
if only i knew... i need to look at all pull requests for 9.4.7 and 9.4.8
fail cabal ci https://github.com/haskell/cabal/actions/runs/15363053137/job/43232541759?pr=10969#step:17:952

zlonast replied:
I still don't know where it was fixed, my search methods were weak. I only know that it happened with the release 9.4.*

zlonast replied:
The solution is here https://github.com/haskell/cabal/pull/11713

zlonast replied:
Now this is merged into master, I'm closing the discussions

ulysses4ever commented:
rebasing on master is also advisable

zlonast commented:
@mpickering
I tried to correct your comments.
True, I don't know how to run stackage build. (But I would be glad if you told me how you did it) (for now it seems scary to try to install all the packages)
Also, I don't know how to comment on "problems with linking", since I'm not sure how exactly it was fixed.

geekosaur commented:
Perhaps https://github.com/haskell/clc-stackage would be of help here?

zlonast commented:
> Perhaps https://github.com/haskell/clc-stackage would be of help here?

Yes thanks.

> I compiled stackage with this branch and GHC-9.10.2 and all the packages compiled.

I'll be honest, I have one error. And because the library has no flags for mac os.

https://hackage.haskell.org/package/clock-0.8.4/docs/System-Clock.html#t:Clock

Constructors | Comments
-- | --
MonotonicRaw | (since Linux 2.6.28, macOS 10.12)
MonotonicCoarse | (since Linux 2.6.32; Linux-specific)

Error with `om-time` package on my mac os:

```
Resolving dependencies...
Build profile: -w ghc-9.10.2 -O0
In order, the following will be built (use -v for more details):
 - om-time-0.3.1.1 (lib) (requires build)
 - generated-0.1.0.0 (lib) (configuration changed)
Starting     om-time-0.3.1.1 (lib)
Building     om-time-0.3.1.1 (lib)

Failed to build om-time-0.3.1.1.
Build log (
/Users/iliabaryshnikov/.cabal/logs/ghc-9.10.2/m-tm-0.3.1.1-ffda054d.log ):
Warning: this is a debug build of cabal-install with assertions enabled.
Configuring library for om-time-0.3.1.1...
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for om-time-0.3.1.1...
Building library for om-time-0.3.1.1...
[1 of 1] Compiling OM.Time          ( src/OM/Time.hs, dist/build/OM/Time.o, dist/build/OM/Time.dyn_o )
src/OM/Time.hs:56:27: error: [GHC-76037]
    Not in scope: data constructor ‘Clock.MonotonicCoarse’
    NB: the module ‘System.Clock’ does not export ‘MonotonicCoarse’.
    Suggested fix:
      Perhaps use ‘Clock.MonotonicRaw’ (imported from System.Clock)
   |
56 |   getTime = Clock.getTime Clock.MonotonicCoarse
   |                           ^^^^^^^^^^^^^^^^^^^^^
```

phadej commented:
I realised that we should only pass `-pgmc gcc` to GHC when user explicitly passed `--with-gcc` to `cabal-install` / `Cabal`. GHC has its own C compiler (in `ghc --info`) and uses it, and we probably shouldn't "overwrite" it if not explicitly asked.

In fact, if we need to run `gcc` explicitly ourselves with GHC (which I think we should not need), then we should ask for `C compiler` command from GHC.

If this insight makes something in this PR easier, go for it; if not, ignore for time being.

Mikolaj commented:
@zlonast: anything blocking this except for the second review?

@mpickering: could you please have another look now?

zlonast commented:
Perhaps we should think about what link-related testing strategy we would like to choose before accepting prs like this 🤔

geekosaur reviewed (approved):
There are a lot of unresolved comments still, that will need to be resolved to merge.

zlonast committed:
Pass *-options to GHC
Refactor componentGhcOptions for pass *-options to all invoking GHC.

During the refactoring process we needed to add componentGhcOptions
to all Haskell sources. It was also worth add linkGhcOptions
to linkLibrary same as componentGhcOptions to linkExecutable, linkFLib.

Add test for PackageTests/FFI/ForeignOptsCapi to pass cc-options flags to *.h.

Fixes #9801 #4435

Co-authored-by: Mikolaj Konarski <mikolaj.konarski@gmail.com>

zlonast committed:
add test for foreign source in TemplateHaskell

Co-authored-by: Yuriy Syrovetskiy <git-orbi@cblp.su>

zlonast commented:
To summarize the current state of this PR:

- All previous discussions have been resolved.
- All prerequisite PRs have been merged.
- Tests are written and passing.

Regarding the tests, the two added test cases directly address what was requested in the linked issues (#9801, #4435):
1. The first test `ForeignOptsCapi` demonstrates that the flags now successfully reach the linking objects.
2. The second test `ForeignSourceTemplateHaskell` shows the flexibility of our approach: everything works as expected even within `TemplateHaskell`.

@Mikolaj @ulysses4ever I believe this is now ready.

ulysses4ever commented on Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs:
```diff
@@ -90,7 +110,26 @@ buildCSources mbMainFile =
+            -- see example in cabal-testsuite/PackageTests/FFI/ForeignOptsPgmc
+            ghcOptCcProgram = maybeToFlag $ programPath <$> lookupProgram gccProgram (withPrograms lbi)
+          }
+    )
```
Is this an absolute duplicate of the hunk above? Can we factor it out? At the very least, I don't think we should duplicate comments that long. You could create a "Note" and reference it. (We don't use notes as extensively as GHC, but there are some cases where we do.)

zlonast replied:
Then I propose what is in the last commit

ulysses4ever commented on Cabal/src/Distribution/Simple/GHCJS.hs:
```diff
@@ -1443,13 +1442,24 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
+                  -- we pass the flag only to source files #11712
+                  -- see example in cabal-testsuite/PackageTests/FFI/ForeignOptsPgmc
+                  ghcOptCcProgram = maybeToFlag $ programPath <$> lookupProgram gccProgram (withPrograms lbi)
+                }
```
Same.

ulysses4ever commented on Cabal/src/Distribution/Simple/GHCJS.hs:
```diff
@@ -1489,13 +1499,24 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
+                  -- we pass the flag only to source files #11712
+                  -- see example in cabal-testsuite/PackageTests/FFI/ForeignOptsPgmc
+                  ghcOptCcProgram = maybeToFlag $ programPath <$> lookupProgram gccProgram (withPrograms lbi)
+                }
```
Same.

ulysses4ever commented on changelog.d/pr-10969.md:
```diff
@@ -0,0 +1,22 @@
+then in 8.10 they started to split into `-optc` and `-optcxx`, cabal picked up
+this change and started to give flags separately not only for new versions,
+but also for old ones. That is, now for 8.8 we send `-optc` flags separately for
+different sources.
```
```suggestion
Before GHC 8.10, Cabal passed its `cc-options` and `cxx-options` via one GHC flag `-optc`.
GHC  8.10 introduced `-optcxx`, and Cabal picked up this change: `cc-options` and `cxx-options` are now mapped to `-optc` and `-optcxx` respectively.
But Cabal also changed its behaviour for older GHCs (8.8 and older): Cabal sends `cc-options` and `cxx-options` through separate instances of `-optc`.
```

Does this sound right? I think it's a bit more clear.

ulysses4ever replied:
Actually, changelog may be not the best place to put historical notes. Can you move it to commit message or/and comments? 

For Changelog, it would be relevant to comment on compatibility with GHCs pre 9.4. Does this patch change anything for thise? I still don't quite understand those comments about 9.4 in the sources.

zlonast replied:
Yes, you're right, I just redid the comment.

```haskell
          -- We pass -pgmc to ensure that GHC respects cc-options and ld-options (#4435, #9801).
          -- However, we deliberately restrict this flag ONLY to C and C++ source files.
          -- We do not pass it globally (e.g., during linking or ordinary Haskell compilation)
          -- for two reasons:
          -- 1. Status quo: This preserves Cabal's historical behavior.
          -- 2. GHC < 9.4 bug (https://gitlab.haskell.org/ghc/ghc/-/issues/15319): Before GHC 9.4,
          --    passing a custom C compiler via -pgmc caused GHC to automatically disable the
          --    -no-pie flag (which is critical during linking on auto-hardened toolchains).
          --    Passing -pgmc globally would therefore break linking on older GHCs.
          --    https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6949 fixed this by moving
          --    the -no-pie check to -pgml, but scoping -pgmc exclusively to C/C++ compilation
          --    protects older GHC versions from this link-time breakage.
```

zlonast committed:
[Historical note]
Before GHC 8.10, Cabal passed both `cc-options` and `cxx-options` via a single GHC flag, `-optc`.
When GHC 8.10 introduced `-optcxx`, Cabal picked up this change: `cc-options` and `cxx-options` are now mapped to `-optc` and `-optcxx`, respectively.
However, Cabal also changed its behavior for older GHC versions (8.8 and below): it now sends `cc-options` and `cxx-options` through separate instances of `-optc`.

Co-authored-by: Artem Pelenitsyn <a.pelenitsyn@gmail.com>

Mikolaj commented:
This looks great! Are the conversations resolved but not marked, or not yet resolved?

zlonast commented:
@ulysses4ever For now, I'll put the `merge me` label, but if you want me to change the comments, feel free to write about it
